osutil/epoll/epoll.go

```diff
@@ -0,0 +1,177 @@
+func (e *Epoll) Close() error {
+	runtime.SetFinalizer(e, nil)
+	e.countMutex.Lock()
+	fd := e.fd
+	e.fd = -1
+	e.registeredFdCount = 0
+	e.countMutex.Unlock()
+	return unix.Close(fd)
+}
+
+// Fd returns the integer unix file descriptor referencing the open file.
+func (e *Epoll) Fd() int {
+	return e.fd
+}
+
+// RegisteredFdCount returns the number of file descriptors which are currently
+// registered to the epoll instance.
+func (e *Epoll) RegisteredFdCount() int {
+	return e.registeredFdCount
+}
+
+// Register registers a file descriptor and allows observing speicifc I/O readiness events.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_ADD for details.
+func (e *Epoll) Register(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_ADD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount += 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Deregister removes the given file descriptor from the epoll instance.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_DEL for details.
+func (e *Epoll) Deregister(fd int) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_DEL, fd, &unix.EpollEvent{})
+	if err != nil {
+		return err
+	}
+	e.countMutex.Lock()
+	e.registeredFdCount -= 1
+	e.countMutex.Unlock()
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Modify changes the set of monitored I/O readiness events of a previously registered file descriptor.
+//
+// Please refer to epoll_ctl(2) and EPOLL_CTL_MOD for details.
+func (e *Epoll) Modify(fd int, mask Readiness) error {
+	err := unix.EpollCtl(e.fd, unix.EPOLL_CTL_MOD, fd, &unix.EpollEvent{
+		Events: uint32(mask),
+		Fd:     int32(fd),
+	})
+	runtime.KeepAlive(e)
+	return err
+}
+
+// Event describes an IO readiness event on a specific file descriptor.
+type Event struct {
+	Fd        int
+	Readiness Readiness
+}
+
+// WaitTimeout blocks and waits with the given timeout for arrival of events on any of the added file descriptors.
+//
+// A msec value of -1 disables timeout.
+//
+// Please refer to epoll_wait(2) and EPOLL_WAIT for details.
+//
+// Warning, using epoll from Golang explicitly is tricky.
+func (e *Epoll) WaitTimeout(duration time.Duration) ([]Event, error) {
+	msec := int(duration.Milliseconds())
+	if duration < 0 {
+		msec = -1
+	}
+	sysEvents := make([]unix.EpollEvent, e.registeredFdCount)
+	n, err := unix.EpollWait(e.fd, sysEvents, msec)
+	runtime.KeepAlive(e)
+	if err != nil {
+		return nil, err
+	}
+	events := make([]Event, 0, n)
+	for i := 0; i < n; i++ {
+		event := Event{
+			Fd:        int(sysEvents[i].Fd),
+			Readiness: Readiness(sysEvents[i].Events),
+		}
+		events = append(events, event)
+	}
+	return events, nil
+}
+
+// Wait blocks and waits for arrival of events on any of the added file descriptors.
+func (e *Epoll) Wait() ([]Event, error) {
+	duration, err := time.ParseDuration("-1ms")
```

Comment:
(nitpick) given that time.Duration is just a int64, this could be written as:
```go
func (e *Epoll) Wait() ([]Event, error) {
	return e.WaitTimeout(time.Duration(-1))
}
```
I think

Reply:
As-is, the -1ns would be rounded down to 0ms, which would not work, but I did add a check for if duration is negative at the beginning of `WaitTimeout()`, so this should be safe to do now instead. I'll make the change, thanks!